### DISCUSSION TITLE
How to use the `after_begin` event since SQLAlchemy 2.0.17?

### DISCUSSION THREAD

**Discussion by pradasouvanlasy** [sqlalchemy/sqlalchemy#10469] [Usage Questions]
Hello!

We got RLS enabled on our Postgres database. Basically, we are relying on the `after_begin` event to configure the relevant variable depending of the current context at the transaction level as follows:
```
@event.listens_for(Session, "after_begin")
def receive_after_begin(session, transaction, connection):
    ... # Specific business unrelated with SQLAlchemy omitted
    statement = text("SET LOCAL ...") # Configures the needed variable at the transaction level for RLS purposes
    session.execute(statement)
```

The `AsyncSession` is then provided through a context manager when needed and follow the same lifecycle as the incoming HTTP request:
```
SessionLocal = async_sessionmaker(
    autocommit=False,
    autoflush=False,
    bind=engine,
    class_=AsyncSession,
    expire_on_commit=False,
)

@asynccontextmanager
async def session_factory():
    db: AsyncSession = SessionLocal()
    try:
        yield db
    finally:
        await db.close()
```

We're aware about `AsyncSession` not being thread-safe and that state checks have been enforced since 2.0.17. We are using sessions as follows since the beginning: 
```
async with session_factory() as session:
    ...
    await session.commit()
```

The first `SET LOCAL ...` statement will be executed as expected but then the next `commit` statement always raises `sqlalchemy.exc.InvalidRequestError: This session is provisioning a new connection; concurrent operations are not permitted` since SQLAlchemy 2.0.17 - the same worked until 2.0.16.

Sorry in advance if we missed something in the documentation but how such `after_begin` hooks are now meant to be implemented if we can no longer interact with the session? Is there another option? Thanks!

**Comment by zzzeek** [comment]
I would assume you mean the 2.0 series in general, I dont think anything changed here in 2.0.17 specifically.

that after_begin() event is handing you the `Connection` object that's the subject of the transaction.  call "SET LOCAL" on *that*, not the `Session`, since `Session` as you observe is in the middle of a state change.

**Reply by pradasouvanlasy** [reply]
Thanks for that answer, will give this a try. I amended the original post since calling `SET LOCAL` on the session worked until 2.0.16 - probably consequently to https://github.com/sqlalchemy/sqlalchemy/commit/70e0f4c0fae1f7410a55c19e1282d4b4a179adc9 that was introduced since 2.0.17.

**Reply by zzzeek** [reply]
oh yes look at that.   OK sure, we added one extra check in that commit, after getting **lots** of folks with fastAPI not using the session correctly.

using the "connection" to emit SQL control statements and such within after_begin() is the approach you'd take for all SQLAlchemy versions; this command does not need to be managed by the Session.

**Reply by pradasouvanlasy** [reply]
That's understandable you added this check and appreciate that very quick answer there. Just tested this simple change and seems to behave as I would expect so far

**Comment by lelit** [comment]
Is there any significant difference between using the [ConnectionEvents.begin](https://docs.sqlalchemy.org/en/20/core/events.html#sqlalchemy.events.ConnectionEvents.begin) and the mentioned [SessionEvents.after_begin](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.SessionEvents.after_begin) to accomplish this?

**Reply by zzzeek** [reply]
for simple cases not really but it's highly different if the Session is bound to a connection directly, or the Session is using savepoints, etc.   The session level after_begin() event corresponds to when the Session itself considers the given connection to be part of its own SessionTransaction state, even if the connection were already in a transaction when passed to the Session (see https://docs.sqlalchemy.org/en/20/orm/session_transaction.html#joining-a-session-into-an-external-transaction-such-as-for-test-suites and https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.params.join_transaction_mode for various cases here) .   session level after_begin() is also called if there's a savepoint (nested) transaction being called on the Session.     All of this can be detected within the session-level after_begin() event by inspecting the SessionTransaction that's passed in.

given that, if the directives here are truly local to the real DBAPI-level transaction then yes, the connection-level after_begin event might be preferable, depending on what state it is that's needed in this command being emitted.

**Reply by lelit** [reply]
Thanks Mike.
